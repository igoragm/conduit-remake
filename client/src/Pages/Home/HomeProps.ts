import IArticleListItem from "Components/Modules/ArticleList/IArticleListItem";
import { ActionFromReducer, Dispatch } from "redux";
import articlesReducer from "../../store/articleList/articleListReducer";
import { HOME_PAGE_LOADED_START, UPDATE_TAG, APPLY_TAG_FILTER_START, SET_LOADING } from "../../constants/actionTypes";

// tslint:disable-next-line:no-empty
export const GetProps = () => {};

export const mapStateToProps = () => (articles: IArticleListItem[], loadingIndicator: boolean) => ({
    articles,
    loadingIndicator
});

export const mapDispatchToProps = () => (dispatch: Dispatch<ActionFromReducer<typeof articlesReducer>>) => ({
    getArticles: () => {
        dispatch({ type: HOME_PAGE_LOADED_START });
    },
    onClickTag: (tag: string, pageManager: () => void) => {
        dispatch({ type: UPDATE_TAG, activeTag: tag });
        dispatch({ type: APPLY_TAG_FILTER_START, tag, pageManager });
    },
    onClickTagByPage: (tag: string, page: number) => {
        dispatch({ type: SET_LOADING });
        dispatch({ type: APPLY_TAG_FILTER_START, tag, page });
    }
});
