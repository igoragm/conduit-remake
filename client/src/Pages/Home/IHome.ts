interface IHome {
  articles: any;
  onClickTag: (tag: string, onClickTagByPage: () => void) => void;
  onClickTagByPage: any;
  getArticles: () => void;
  loading: boolean;
}

export default IHome;
