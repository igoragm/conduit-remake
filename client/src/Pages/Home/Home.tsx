/* eslint-disable react-hooks/exhaustive-deps */
import React, { FC, Fragment } from "react";
import { Banner } from "Components/Modules/Banner/Banner";
import MainView from "Components/Modules/MainView/MainView";
import Tags from "Components/Modules/Tags";
import IHome from "./IHome";
import styles from "./Home.module.scss";
import { mapStateToProps, mapDispatchToProps } from "./HomeProps";
import { connect } from "react-redux";

export const Home: FC<IHome> = (props: IHome) => {
    React.useEffect(() => {
        props.getArticles();
    }, []);

    return (
        <Fragment>
            <Banner appName="conduit app remake" />

            <div data-testid="home" className={styles.splitscreen}>
                <MainView className={styles.left} {...props.articles.articleListReducer} />

                <div className={styles.right}>
                    <>
                        <p>Popular Tags</p>

                        <Tags
                            tags={props.articles.articleListReducer.articles ? props.articles.articleListReducer.articles.tags : []}
                            onClickTag={props.onClickTag}
                            onClickTagByPage={props.onClickTagByPage}
                        />
                    </>
                </div>
            </div>
        </Fragment>
    );
};

export default connect(mapStateToProps(), mapDispatchToProps())(Home);

// https://medium.com/javascript-in-plain-english/i-tested-a-react-app-with-jest-testing-library-and-cypress-here-are-the-differences-3192eae03850
