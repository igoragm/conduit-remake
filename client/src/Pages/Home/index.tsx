import { Home } from "./Home";
import { SetState } from "../../Utils/StateHandler";
import * as GetProps from "./HomeProps";

export default SetState(Home, GetProps);
