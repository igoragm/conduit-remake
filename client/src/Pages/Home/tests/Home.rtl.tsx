import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import { Home } from "../Home";
import IHome from "../IHome";

describe("<Home />", () => {
    it("renders the Home page", async () => {
        const props: IHome = {
            getArticles: jest.fn(),
            articles: {
                articleListReducer: {}
            },
            onClickTag: jest.fn(),
            loading: false,
            onClickTagByPage: jest.fn()
        };

        const homeContainer = render(<Home {...props} />);

        expect(homeContainer).toMatchSnapshot();

        expect(document.querySelector("#banner")?.innerHTML).toBe("conduit app remake");
    });
});

// great tutorial for RTL with TS
// https://www.pluralsight.com/guides/how-to-test-react-components-in-typescript
// for css scss modules use "identity-obj-proxy" for mock in jest config moduleNameMapper npm i identity-obj-proxy -D
// or maybe:
// "transform": {
//     "^.+\\.js$": "babel-jest", - ts-jest in my case
//     ".+\\.(css|styl|less|sass|scss)$": "jest-transform-css"
//   }
