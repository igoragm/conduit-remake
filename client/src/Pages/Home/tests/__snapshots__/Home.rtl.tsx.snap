// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> renders the Home page 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="banner"
      >
        <h1
          class="logoFont"
          id="banner"
        >
          conduit app remake
        </h1>
        <p>
          A place to share your knowledge.
        </p>
      </div>
      <div
        class="splitscreen"
      >
        <div
          class="homeContainer"
        >
          <ul
            class="globalFeedOutline"
          >
            <li
              class="nav-item"
            >
              <a
                class="nav-link active"
                href=""
              >
                Global Feed
              </a>
            </li>
          </ul>
          <div>
            <div
              class="makeStyles-loader-1"
            >
              <div
                class="sc-bdVaJa gfLNMF"
                style="margin: auto; position: absolute; left: 0px; right: 0px; top: 0px; bottom: 0px;"
              >
                <div
                  class="sc-bwzfXH sc-htpNat flDhnP"
                >
                  <span
                    class="sc-bZQynM iepmoa"
                  />
                </div>
                <div
                  class="sc-bwzfXH sc-bxivhb kINVDU"
                >
                  <span
                    class="sc-bZQynM iepmoa"
                  />
                </div>
                <div
                  class="sc-bwzfXH sc-ifAKCX gJwEBr"
                >
                  <span
                    class="sc-bZQynM iepmoa"
                  />
                </div>
                <div
                  class="sc-bwzfXH sc-EHOje hkQVFp"
                >
                  <span
                    class="sc-bZQynM iepmoa"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="right"
        >
          <p>
            Popular Tags
          </p>
          <div
            class="makeStyles-loader-2"
          >
            <div
              class="sc-bdVaJa gfLNMF"
              style="margin: auto; position: absolute; left: 0px; right: 0px; top: 0px; bottom: 0px;"
            >
              <div
                class="sc-bwzfXH sc-htpNat flDhnP"
              >
                <span
                  class="sc-bZQynM iepmoa"
                />
              </div>
              <div
                class="sc-bwzfXH sc-bxivhb kINVDU"
              >
                <span
                  class="sc-bZQynM iepmoa"
                />
              </div>
              <div
                class="sc-bwzfXH sc-ifAKCX gJwEBr"
              >
                <span
                  class="sc-bZQynM iepmoa"
                />
              </div>
              <div
                class="sc-bwzfXH sc-EHOje hkQVFp"
              >
                <span
                  class="sc-bZQynM iepmoa"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="banner"
    >
      <h1
        class="logoFont"
        id="banner"
      >
        conduit app remake
      </h1>
      <p>
        A place to share your knowledge.
      </p>
    </div>
    <div
      class="splitscreen"
    >
      <div
        class="homeContainer"
      >
        <ul
          class="globalFeedOutline"
        >
          <li
            class="nav-item"
          >
            <a
              class="nav-link active"
              href=""
            >
              Global Feed
            </a>
          </li>
        </ul>
        <div>
          <div
            class="makeStyles-loader-1"
          >
            <div
              class="sc-bdVaJa gfLNMF"
              style="margin: auto; position: absolute; left: 0px; right: 0px; top: 0px; bottom: 0px;"
            >
              <div
                class="sc-bwzfXH sc-htpNat flDhnP"
              >
                <span
                  class="sc-bZQynM iepmoa"
                />
              </div>
              <div
                class="sc-bwzfXH sc-bxivhb kINVDU"
              >
                <span
                  class="sc-bZQynM iepmoa"
                />
              </div>
              <div
                class="sc-bwzfXH sc-ifAKCX gJwEBr"
              >
                <span
                  class="sc-bZQynM iepmoa"
                />
              </div>
              <div
                class="sc-bwzfXH sc-EHOje hkQVFp"
              >
                <span
                  class="sc-bZQynM iepmoa"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="right"
      >
        <p>
          Popular Tags
        </p>
        <div
          class="makeStyles-loader-2"
        >
          <div
            class="sc-bdVaJa gfLNMF"
            style="margin: auto; position: absolute; left: 0px; right: 0px; top: 0px; bottom: 0px;"
          >
            <div
              class="sc-bwzfXH sc-htpNat flDhnP"
            >
              <span
                class="sc-bZQynM iepmoa"
              />
            </div>
            <div
              class="sc-bwzfXH sc-bxivhb kINVDU"
            >
              <span
                class="sc-bZQynM iepmoa"
              />
            </div>
            <div
              class="sc-bwzfXH sc-ifAKCX gJwEBr"
            >
              <span
                class="sc-bZQynM iepmoa"
              />
            </div>
            <div
              class="sc-bwzfXH sc-EHOje hkQVFp"
            >
              <span
                class="sc-bZQynM iepmoa"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
